torch/_inductor/index_propagation.py

```diff
@@ -0,0 +1,167 @@
+"""
+This file implements the IndexPropagation ops handler, which wraps an
+underlying handler to add a limited for of constant propagation, as well as
+propagation of sympy expressions downstream of `ops.index_expr` calls.
+
+"""
+import itertools
+from dataclasses import dataclass
+from typing import Any, Optional
+
+import sympy
+
+import torch
+from torch._prims_common import is_boolean_dtype, is_integer_dtype
+
+
+@dataclass
+class TypedExpr:
+    """A SymPy expression with associated type"""
+
+    expr: sympy.Expr
+    dtype: torch.dtype
+
+
+class SymPyOps:
+    """An ops handler where all IR values are SymPy expressions
+
+    When a value cannot be represented as a SymPy expression, the method is
+    either not defined, or returns NotImplemented
+
+    """
+
+    @staticmethod
+    def identity(value):
+        return value
+
+    @staticmethod
+    def constant(value, dtype):
+        if is_boolean_dtype(dtype):
+            expr = sympy.Integer(bool(value))
+        elif is_integer_dtype(dtype):
+            expr = sympy.Integer(int(value))
+        else:
+            expr = sympy.Float(float(value))
+        return TypedExpr(expr, dtype)
+
+    @staticmethod
+    def index_expr(value, dtype):
+        if isinstance(value, int):
+            value = sympy.Integer(value)
+        return TypedExpr(value, dtype)
+
+    @staticmethod
+    def to_dtype(value, dtype):
+        if isinstance(value.expr, (sympy.Integer, sympy.Float)):
+            return SymPyOps.constant(value.expr, dtype)
+        elif is_integer_dtype(dtype) and is_integer_dtype(value.dtype):
+            return SymPyOps.index_expr(value.expr, dtype)
+        else:
+            # TODO: Inductor doesn't handle floating point in sympy expressions well at the moment
+            return NotImplemented
+
+    @staticmethod
+    def square(x):
+        return TypedExpr(x.expr * x.expr, x.dtype)
+
+    @staticmethod
+    def add(x, y):
+        result_type = torch.promote_types(x.dtype, y.dtype)
+        return TypedExpr(x.expr + y.expr, result_type)
+
+    @staticmethod
+    def sub(x, y):
+        result_type = torch.promote_types(x.dtype, y.dtype)
+        return TypedExpr(x.expr - y.expr, result_type)
+
+    @staticmethod
+    def mul(x, y):
+        result_type = torch.promote_types(x.dtype, y.dtype)
+        return TypedExpr(x.expr * y.expr, result_type)
+
+    @staticmethod
+    def neg(x):
+        return TypedExpr(-x.expr, x.dtype)
+
+
+@dataclass
+class IndexPropVar:
+    rep: Any  # A wrapped IR value, if present
+    expr: Optional[TypedExpr] = None
```

Comment:
Could this be written as
```suggestion
@dataclass
class IndexPropVar:
    rep: Union[TypedExpr, Any] = None  # Any == wrapped IR value
```
?

Reply:
In the unlikely event that the wrapped handler returns `TypedExpr`, this would behave unexpectedly.

Reply:
Then let's assert that it doesn't do that and change it to the more compact representation? I believe it's makes easier to understand what's really going on (we can get rid of the `id_iter` and the `index_tmp` name that's never used).